classes/services/PKPFileService.inc.php

```diff
@@ -0,0 +1,251 @@
+<?php
+/**
+ * @file classes/services/PKPFileService.php
+ *
+ * Copyright (c) 2014-2019 Simon Fraser University
+ * Copyright (c) 2000-2019 John Willinsky
+ * Distributed under the GNU GPL v2. For full terms see the file docs/COPYING.
+ *
+ * @class PKPFileService
+ * @ingroup services
+ *
+ * @brief Helper class that encapsulates business logic for publications
+ */
+
+namespace PKP\Services;
+
+use \Config;
+use \Exception;
+use \HookRegistry;
+use League\Flysystem\Adapter\Local;
+use League\Flysystem\Filesystem;
+use Illuminate\Database\Capsule\Manager as Capsule;
+
+class PKPFileService {
+
+	/** @var Filesystem */
+	public $fs;
+
+	/**
+	 * Initialize and configure flysystem
+	 */
+	public function __construct() {
+
+		$adapter = new Local(
+			Config::getVar('files', 'files_dir'),
+			LOCK_EX,
+			Local::DISALLOW_LINKS,
+			[
+				'file' => [
+					'public' => FILE_MODE_MASK,
+					'private' => FILE_MODE_MASK,
+				],
+				'dir' => [
+					'public' => DIRECTORY_MODE_MASK,
+					'private' => DIRECTORY_MODE_MASK,
+				]
+			]
+		);
+
+		HookRegistry::call('File::adapter', [$adapter, $this]);
+
+		$this->fs = new Filesystem($adapter);
+	}
+
+	/**
+	 * Get a file path by its id
+	 *
+	 * @param int $id
+	 * @return string
+	 */
+	public function getPath($id) {
+		$file = Capsule::table('files')
+			->where('file_id', '=', $id)
+			->first();
+		return $file ? $file->path : '';
+	}
+
+	/**
+	 * Add a file
+	 *
+	 * @param string $from
+	 * @param string $to
+	 * @return int file id
+	 */
+	public function add($from, $to) {
```

Comment:
Why not just use FlySystem's `copy` function?

Reply:
I can't pretend to know what I'm talking about here, but I think I thought that the `writeStream` method was more performant for large files, especially with remote file systems? Happy to use `copy()` here.

Reply:
Using the `copy` method won't work because Flysystem expects the file paths to be relative to its root directory, which is OJS's `files_dir`. In this case, we're usually writing from the system's `tmp` directory, so the `$from` path can not be used with `copy`.

I've reverted to the existing code but let me know if we should change it in some way.

Reply:
:+1: OK, thanks for looking into it!